src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+section \<open>Chain Synchronization Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+subsection \<open>Cryptographic Primitives\<close>
+
+typedecl hash
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
+
+subsection \<open>Chains\<close>
+
+type_synonym chain = "block list"
+
+type_synonym chain_index = nat
+
+abbreviation (input) tip_index :: "chain \<Rightarrow> chain_index" where
+  "tip_index \<C> \<equiv> length \<C> - 1"
+
+abbreviation (input) tip :: "chain \<Rightarrow> block" where
+  "tip \<C> \<equiv> \<C> ! (tip_index \<C>)"
```

Comment:
Generally, we should avoid `type_synonym` and `abbreviation`, which do not introduce new entities from the point of the logic, and rather introduce dedicated types (via `datatype`, `record`, etc.) and constants (via `definition`, `primrec`, etc.).

I used to use `abbreviation` instead of `definition` long ago but then turned away from this habit, actually because *you* had convinced me to do so :wink:. These days, I’m using `abbreviation` only for introducing alternative notation with swapped arguments, like the infix notation of the conformance relation (`conformance_std`).

Reply:
That said, I have also some remarks that are specific to this code snippet.

First, introducing `chain_index` at all seems overkill to me. An index is almost always a natural number; so it would make sense to use `nat` directly.

Second, I also suppose that the introduction of `tip_index` and `tip` doesn’t really help comprehensibility but may rather make things more complicated to understand. Dropping `tip_index` might still be debatable, as there doesn’t seem to be a concise notation of length, like `|_|`, but `tip` is just the same as the `last` function from the standard library, isn’t it?

Reply:
That's right. Probably it would be better to drop `tip_index` and define `tip` as `last`.

Reply:
I wouldn’t define `tip` at all but rather use `last` where currently `tip` is used. If it’s really the same function, it’s maybe misleading to use a dedicated function like `tip`, because it suggests that there is something special about it, which actually isn’t.